docs/topics/coroutines.rst

```diff
@@ -42,10 +42,71 @@ hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
     methods of
     :ref:`downloader middlewares <topics-downloader-middleware-custom>`.
 
+-   The
+    :meth:`~scrapy.spidermiddlewares.SpiderMiddleware.process_start_requests`
+    method of :ref:`spider middlewares <custom-spider-middleware>`.
+
 -   :ref:`Signal handlers that support deferreds <signal-deferred>`.
 
+-   The :meth:`~scrapy.spiders.Spider.start_requests` spider method.
+
 .. _asynchronous generators were introduced in Python 3.6: https://www.python.org/dev/peps/pep-0525/
 
+.. _async-start_requests:
+
+Asynchronous start_requests and spider middlewares
+==================================================
+
+.. versionadded:: 2.1
+
+The :meth:`~scrapy.spiders.Spider.start_requests` spider method can be an
+asynchronous generator. In this case all spider middlewares used with this
+spider that have the
+:meth:`~scrapy.spidermiddlewares.SpiderMiddleware.process_start_requests`
+method must support this: if they receive an asynchronous iterable, they must
+return one as well. On the other hand, if they receive a normal iterable, they
+shouldn't break and ideally should return a normal iterable too. There can be
+several possible implementations of this.
+
+First, such universal :meth:`process_start_requests` can be an asynchronous
+generator itself, and so it will always convert a normal iterable to an
+asynchronous one. Because a result of a middleware method is passed to the same
+method of the next middleware, it's only possible to mix middlewares with
+synchronous and asynchronous :meth:`process_start_requests` if all synchronous
+ones are called first.
+
+.. autofunction:: scrapy.utils.asyncgen.as_async_generator
+
+Here is an example of a universal middleware using this approach::
+
+    from scrapy.utils.asyncgen import as_async_generator
+
+    class ProcessStartRequestsAsyncGenMiddleware:
+        async def process_start_requests(self, start_requests, spider):
+            async for r in as_async_generator(start_requests):
```

Comment:
Could you please change r to req or request?

Reply:
Also, whatdo you think about making process_start_requests async in the gerenated templates?

Reply:
@kmike do you mean https://github.com/scrapy/scrapy/blob/master/scrapy/templates/project/module/middlewares.py.tmpl ?

Reply:
yes

Reply:
I don't have an opinion on this

Reply:
What's the downside? I can see 3 downsides:

1. Generated code becomes more complicated
2. Generated code won't work with Python 3.5
3. We'll be changing the recommended way to write middleware in future, making this generated code deprecated.

I think (2) and (3) could be valid reasons not to have it as an example. Are they the case? 
As for (1), I think it worths it to have a proper implementation there, even if it means slightly more generated code.

Reply:
It's more of "do we need to change this if most people will still write only normal start_requests and normal middlewares for them". It also can be more complicated for people not familiar with async def functions, but that's probably not a big problem. 

About (2): assuming most of the use cases need an iteration, straightforward `for` loop with `yield` will indeed not work in 3.5. About (3): I'm still not sure what should we recommend for universal middlewares, but we can always update the template again.